### PR TITLE
Allow files generated during build to be used as sources.

### PR DESCRIPTION
**Pull request by TamarMilchtaich**
### Short description 📝

There are cases where it's useful to use the output of a pre-build phase script as a resource to the built target. This PR supports it by allowing such generated files to be used.

Resolves <https://github.com/tuist/tuist/issues/6602>

### How to test the changes locally 🧐

You can test it with files generated during a build phase scrip.t Use a project with the following:
1. A pre- target script with output file to use in the project. You can synthisize one by using a shell command in the lines of `touch "example.swift"` as your script, or adding an `output file` in the script and printing to it.
2. Add the script generated to your sources in this target, e.g.:
```
  sources: [
      .generated("myGeneratedFile.swift")
  ],
```
You can try it both when the generated file is in the Derived Data folder and when it's not.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (fortmarek) [inline Sources/ProjectDescription/SourceFilesList.swift:79]: What do you think of naming this `generated` only? I understand the main purpose here is for scripts, but from Tuist's perspective, it doesn't really matter how exactly these files are generated – what's important is that they are always included in the generated Xcode project.

If so, then `GenerationSource` should also be refactored to something like:
```swift
public var generated: Bool
```

or:
```swift
public var type: SourceFileType

enum SourceFileType {
   /// Always present on the disk before generating a project
  case alwaysPresent
  /// Might not be present when generating a project and can be generated later on
  case generated
```

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (fortmarek) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:289]: I'd move this in the initial [map](https://github.com/tuist/tuist/pull/6990/files#diff-26a125b6de2c8c5bef2552bd263ab117d3aa62a6a6cd69c7e4ba045f48fecca1R257) and create appropriate helper functions instead if that `map` gets too complex.

**Review by fortmarek — COMMENTED** [review]
Thanks for the PR! A couple suggestions 🤞

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (TamarMilchtaich) [reply Sources/ProjectDescription/SourceFilesList.swift:79]: My thinking was that if we know it came from scripts, we could in the future do things like verify that it's listed as an output path of some script. And if there will be another source of generation that's not from script, there could be special treatment for this case as well.

So I think that question is - do we anticipate other sources in the future. But I think that moving to `SourceFileType` can allow adding other source in the future while making the naming clearer, so we can enjoy both.

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (TamarMilchtaich) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:289]: After trying - since they have a different return type (the former returns `SourceFileGlob` and the latter `SourceFile`, which is the type returned from `XcodeGraph.Target.sources` in the original map), attempting to share the code creates more code than it saves, so keeping as is for now. If this is not what you meant or you have a different suggestion - PLMK.

Comment (fortmarek) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift]: I'd do the following instead:
```swift
sources: manifest.sources?.globs
   .map {
     switch $0 {
       case .alwaysPresent:
        return alwaysPresentSourceFiles(...)
      case .generated:
       return generatedSourceFile(...)
     }
   }
```

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (fortmarek) [inline Sources/ProjectDescription/SourceFilesList.swift:7]:
```suggestion
        /// File is already present on disk before generating the project.
```

**Review by fortmarek — COMMENTED** [review]
Can you also add a fixture that showcases the fixture along with an acceptance test? 🙏

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (TamarMilchtaich) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift]: Note that generated sources that are sent to `XcodeGraph.Target.sources(...)` method will eventually be filtered out (by `validateGlobPattern`), so calling `generatedSourceFile(...)` replaces the call to `XcodeGraph.Target.sources(...)`. There's still a possibility to do something like:
```
        manifest.sources?.globs.forEach { glob in
            if glob.type == .generated {
                generatedSource.append(glob)
            }
            else if glob.type == .alwaysPresent {
                presentSources.append(glob)
            }
        }
```
And then use each source array in the corresponding method

**Comment by TamarMilchtaich** [comment]
Fixed up to pending comment (fixture added) + marked as ready for review

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (fortmarek) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:20]:
```suggestion
            return "Generated source files must be explicit. The target \(targetName) has a generated source file at \(generatedSource.pathString) that has a glob pattern."
```

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (fortmarek) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift]: Ahh, my bad, I understand the issue now – let's keep as-is

**Review by fortmarek — DISMISSED** [review]
Looks good! I'd wait for one extra review from @tuist/community-core to sign off on the `ProjectDescription` changes.

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (danieleformichelli) [inline Sources/ProjectDescription/SourceFilesList.swift:12]: I'm a bit worried about target hashing, as we has no info here to determine if the generated file will change or not.
Should we at least allow here a list of files which hare "inputs" to the generation, so we can use them in hashing?
Something like:
```
case generated(inputs: …)
```

**Review by pepicrft — DISMISSED** [review]
API-wise it looks to me. This has implications in caching, which assumes the presence of hashable files ahead of generating the project.
We can always expose an API to extend the hashing logic, but that'll only work as long as the generated files don't depend on build-time information.

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (fortmarek) [reply Sources/ProjectDescription/SourceFilesList.swift:12]: Yeah, that's a good point.

When the file is generated using a build phase, the input and output paths should already be in the final hash.

Comment (danieleformichelli) [reply Sources/ProjectDescription/SourceFilesList.swift:12]: Makes sense, but I would add a disclaimer that those should be set, as IIRC they are optional in the build phase

**Review by danieleformichelli — DISMISSED** [review]

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (fortmarek) [reply Sources/ProjectDescription/SourceFilesList.swift:12]: Doesn't Xcode complain if you don't set them?

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (TamarMilchtaich) [reply Sources/ProjectDescription/SourceFilesList.swift]: Thanks for noticing this mistake! Fixed the phrasing

**Comment by TamarMilchtaich** [comment]
New commit pushed with the changes discussed, PLMK if anything is still missing/should be changed

File: Sources/ProjectDescription/SourceFilesList.swift
Comment (fortmarek) [inline Sources/ProjectDescription/SourceFilesList.swift:13]:
```suggestion
        /// - Important: Since generated files do not exist at the time of project generation, their content cannot be hashed,
```

**Review by fortmarek — APPROVED** [review]

**Pull request merged**
